### ISSUE TITLE
CA1812 false positive on internal static class with only private members

### ISSUE THREAD

**Issue by thomaslevesque** [dotnet/roslyn-analyzers#4677]
### Analyzer

**Diagnostic ID**: [CA1812](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1812): `Avoid uninstantiated internal classes`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 5.0.101](https://dotnet.microsoft.com/download/dotnet/5.0)

### Describe the bug

I'm getting the following warning for an internal static class that contains only private members (I know it seems useless, but there's actually a good reason I won't go into here).

> warning CA1812: Foo is an internal class that is apparently never instantiated. If so, remove the code from the assembly. If this class is intended to contain only static members, make it static (Shared in Visual Basic)

It can't be instantiated, since it's static. And the message suggests to make it static, which it already is.

The warning disappears if the class has at least one public member.

### Steps To Reproduce

Write the following code:

```csharp
internal static class Foo
{
    public static void Bar()
    {
    }
}
```

### Expected behavior

No warning

### Actual behavior

> warning CA1812: Foo is an internal class that is apparently never instantiated. If so, remove the code from the assembly. If this class is intended to contain only static members, make it static (Shared in Visual Basic)

## Additional context

N/A

**Comment by mavasani** [comment]
Tagging @Evangelink - this might have been fixed recently. Can you please verify?

**Comment by Youssef1313** [comment]
https://github.com/dotnet/roslyn-analyzers/pull/4395 excluded static classes from analysis. Most likely this is already fixed.

**Comment by thomaslevesque** [comment]
Where can I get this fix? Do I have to install the package? Or will it be included in the .NET 5 SDK?

**Comment by mavasani** [comment]
Please install the 5.0.3 version of the NetAnalyzers NuGet package.

**Comment by Evangelink** [comment]
I cannot reproduce with your example even on FxCop 3.8.0, could you confirm that the code sample is correct and/or using NetAnalyzers 5.0.3 fixes the issue?

**Comment by thomaslevesque** [comment]
> I cannot reproduce with your example even on FxCop 3.8.0, could you confirm that the code sample is correct and/or using NetAnalyzers 5.0.3 fixes the issue?

I can reproduce it using the SDK's built-in analyzers, or with the NetAnalyzers 5.0.0 package, with `<AnalysisMode>AllEnabledByDefault</AnalysisMode>`. I no longer reproduce it with NetAnalyzers 5.0.1 or 5.0.3.

So I guess the problem is fixed, just not in the SDK built-in analyzers yet... Will the fix be included in a future update of the SDK?

**Comment by mavasani** [comment]
> So I guess the problem is fixed, just not in the SDK built-in analyzers yet... Will the fix be included in a future update of the SDK?

It will be available in the .NET 6.0 SDK. It will not be available with patch releases/updates for .NET 5.0 SDK, because this analyzer is not a warning by default, hence will not meet the bar.
